ludwig/benchmarking/reporting.py

```diff
@@ -33,3 +39,92 @@ def create_metrics_report(experiment_name: str) -> Tuple[Dict[str, Any], str]:
     merged_file_path = os.path.join(os.getcwd(), experiment_name, "metrics_report", "{}.json".format("full_report"))
     save_json(merged_file_path, full_report)
     return full_report, merged_file_path
+
+
+def initialize_stats_dict(function_events):
+    """Initialize dictionary which stores resource usage information per tagged code block."""
+    info = dict()
+    for event_name in [evt.name for evt in function_events if "ludwig" in evt.name]:
+        info[event_name] = {"code_block_tag": event_name}
+        info[event_name]["runs"] = []
+    return info
+
+
+def get_memory_details(event):
+    """Get device name and number of bytes (de)allocated during an event."""
+    if event.device_type() in [DeviceType.CPU, DeviceType.MKLDNN, DeviceType.IDEEP]:
+        return "cpu", event.nbytes()
+    elif event.device_type() in [DeviceType.CUDA, DeviceType.HIP]:
+        return f"cuda_{event.device_index()}", event.nbytes()
```

Comment:
What's the expected behavior when the `device_type` is none of these 4?

If `None` is a valid return, the function should be documented as so. If only these 4 device types are valid, then an error should be raised, or we should define a reasonable default, i.e. `(f"custom_device_type_{event.device_index()}", event.nbytes())`